include/mango_v3.hpp

```diff
@@ -197,6 +203,247 @@ struct EventQueue {
+    uint64_t leafCount;
+    AnyNode nodes[BOOK_SIZE];
+
+    struct iterator {
+      Side side;
+      const BookSideRaw &bookSide;
+      std::stack<uint32_t> stack;
+      uint32_t left, right;
+
+      iterator(Side side, const BookSideRaw &bookSide)
+          : side(side), bookSide(bookSide) {
+        stack.push(bookSide.rootNode);
+        left = side == Side::Buy ? 1 : 0;
+        right = side == Side::Buy ? 0 : 1;
+      }
+
+      bool operator==(const iterator &other) const {
+        return &bookSide == &other.bookSide && stack.top() == other.stack.top();
+      }
+
+      iterator &operator++() {
+        if (stack.size() > 0) {
+          const auto &elem = **this;
+          stack.pop();
+
+          if (elem.tag == NodeType::InnerNode) {
+            const auto innerNode =
+                reinterpret_cast<const struct InnerNode *>(&elem);
+            stack.push(innerNode->children[right]);
+            stack.push(innerNode->children[left]);
+          }
+        }
+        return *this;
+      }
+
+      const AnyNode &operator*() const { return bookSide.nodes[stack.top()]; }
+    };
+  };
+
+  const Side side;
+  std::shared_ptr<BookSideRaw> raw = std::make_shared<BookSideRaw>();
+  std::shared_ptr<std::vector<Order>> orders =
+      std::make_shared<std::vector<Order>>();
+};
+
+class Trades {
+ public:
+  auto getLastTrade() const { return latestTrade; }
+
+  bool handleMsg(const nlohmann::json &msg) {
+    // ignore subscription confirmation
+    const auto itResult = msg.find("result");
+    if (itResult != msg.end()) {
+      return false;
+    }
+
+    // all other messages are event queue updates
+    const std::string method = msg["method"];
+    const int subscription = msg["params"]["subscription"];
+    const int slot = msg["params"]["result"]["context"]["slot"];
+    const std::string data = msg["params"]["result"]["value"]["data"][0];
+
+    const auto decoded = solana::b64decode(data);
+    const auto events = reinterpret_cast<const EventQueue *>(decoded.data());
+    const auto seqNumDiff = events->header.seqNum - lastSeqNum;
+    const auto lastSlot =
+        (events->header.head + events->header.count) % EVENT_QUEUE_SIZE;
+
+    bool gotLatest = false;
+    if (events->header.seqNum > lastSeqNum) {
+      for (int offset = seqNumDiff; offset > 0; --offset) {
+        const auto slot =
+            (lastSlot - offset + EVENT_QUEUE_SIZE) % EVENT_QUEUE_SIZE;
+        const auto &event = events->items[slot];
+
+        if (event.eventType == EventType::Fill) {
+          const auto &fill = (FillEvent &)event;
+          latestTrade = std::make_shared<uint64_t>(fill.price);
+          gotLatest = true;
+        }
+        // no break; let's iterate to the last fill to get the latest fill order
+      }
+    }
+
+    lastSeqNum = events->header.seqNum;
+    return gotLatest;
+  }
+
+ private:
+  uint64_t lastSeqNum = INT_MAX;
+  std::shared_ptr<uint64_t> latestTrade = std::make_shared<uint64_t>(0);
```

Comment:
why not just an uint64_t, trade at price 0 is impossible after all

Reply:
will need a mtx/atomic otherwise. same reason why we replaced all mtx'es with shared ptrs. 
event based synchronization between getLastTrade and handleMsg.
can change to nullptr though so client of getLastTrade could parse the validity of returned value.